File: zeroinstall/cmd/show.py
"""
The B{0install show} command-line interface.
"""

# Copyright (C) 2012, Thomas Leonard
# See the README file for details, or visit http://0install.net.

from __future__ import print_function

import os

from zeroinstall import _, SafeException
from zeroinstall.cmd import select, UsageError
from zeroinstall.injector import qdom, selections

syntax = "APP | SELECTIONS"

def add_options(parser):
	parser.add_option("-r", "--root-uri", help=_("display just the root interface URI"), action='store_true')
	parser.add_option("", "--xml", help=_("print selections as XML"), action='store_true')

def handle(config, options, args):
	if len(args) != 1:
		raise UsageError()

	app = config.app_mgr.lookup_app(args[0], missing_ok = True)
	if app is not None:
		sels = app.get_selections()

		r = app.get_requirements()

		if r.extra_restrictions and not options.xml:
			print("User-provided restrictions in force:")
			for uri, expr in r.extra_restrictions.items():
				print("  {uri}: {expr}".format(uri = uri, expr = expr))
			print()
	elif os.path.exists(args[0]):
		with open(args[0], 'rb') as stream:
			sels = selections.Selections(qdom.parse(stream))
	else:
		raise SafeException(_("Neither an app nor a file: '%s'") % args[0])

	if options.root_uri:
		print(sels.interface)
	elif options.xml:
		select.show_xml(sels)
	else:
		select.show_human(sels, config.stores)

def complete(completion, args, cword):
	"""@type completion: L{zeroinstall.cmd._Completion}
	@type args: [str]
	@type cword: int"""
	if len(args) != 1: return
	completion.expand_apps()
	completion.expand_files()

File: ocaml/cli/show.ml
(* Copyright (C) 2013, Thomas Leonard
 * See the README file for details, or visit http://0install.net.
 *)

(** The "0install show" command *)

open Zeroinstall.General
open Support
open Options
module Qdom = Support.Qdom
module Selections = Zeroinstall.Selections
module Apps = Zeroinstall.Apps

let show_human config f sels =
  Format.fprintf f "%a@." (Zeroinstall.Tree.print config) sels

let show_xml sels =
  let out = Xmlm.make_output @@ `Channel stdout in
  Qdom.reindent sels |> Qdom.output out;
  output_string stdout "\n"

let pp_restrictions f =
  XString.Map.iter @@ fun iface expr ->
  Format.fprintf f "@,%s: %s" iface expr

let show_restrictions f r =
  let open Zeroinstall.Requirements in
  if r.extra_restrictions <> XString.Map.empty then (
    Format.fprintf f
      "@[<v2>User-provided restrictions in force:%a@]@.@."
      pp_restrictions r.extra_restrictions
  )

let handle options flags args =
  let s_root = ref false in
  let s_xml = ref false in

  Support.Argparse.iter_options flags (function
    | #common_option as o -> Common_options.process_common_option options o
    | `ShowRoot -> s_root := true
    | `ShowXML -> s_xml := true
  );

  let config = options.config in
  let system = config.system in
  match args with
  | [arg] -> (
      let sels = match Apps.lookup_app config arg with
      | Some app_path ->
          let r = Apps.get_requirements system app_path in
          if not !s_xml && not !s_root then
            show_restrictions options.stdout r;
          Apps.get_selections_no_updates system app_path
      | None ->
          Selections.load_selections config.system arg in

      match (!s_root, !s_xml) with
      | (true, false) -> Format.fprintf options.stdout "%s@." Selections.((root_role sels).iface)
      | (false, true) -> show_xml (Selections.as_xml sels)
      | (false, false) -> show_human config options.stdout sels
      | (true, true) -> Safe_exn.failf "Can't use --xml with --root"
  )
  | _ -> raise (Support.Argparse.Usage_error 1)
